rust-spec/lightclient/supervisor/supervisor_001_draft.md

```diff
@@ -0,0 +1,640 @@
+```
+
+- Implementation remark
+    - none
+- Expected precondition
+    - none
+- Expected postcondition
+    - lightBlock.Header.Height =  genesisDoc.InitialHeight
+    - lightBlock.Header.Time = genesisDoc.GenesisTime
+    - lightBlock.Header.LastBlockID = nil
+    - lightBlock.Header.LastCommit = nil
+    - lightBlock.Header.Validators = genesisDoc.Validators
+    - lightBlock.Header.NextValidators = genesisDoc.Validators
+    - lightBlock.Header.Data = nil
+    - lightBlock.Header.AppState =  genesisDoc.AppState
+    - lightBlock.Header.LastResult = nil
+    - lightBlock.Commit = nil
+    - lightBlock.Validators = genesisDoc.Validators
+    - lightBlock.NextValidators = genesisDoc.Validators
+    - lightBlock.Provider = nil
+- Error condition
+    - none
+
+----
+
+### Configuration Parameters
+
+
+#### **[LC-INV-ROOT-AGREED.1]**
+
+In the Sequential-Supervisor, it is always the case that the primary
+and all secondaries agree on lightStore.Latest(). 
+
+### Assumptions
+
+We have to assume that the initialization data (the lightblock or the
+genesis file) are consistent with the blockchain. This is subjective
+initialization and it cannot be checked locally.
+
+### Invariants
+
+#### **[LC-INV-PEERLIST.1]:**
+
+The peer list contains a primary and a secondary.
+
+> If the invariant is violated, the light client does not have enough
+> peers to download headers from. As a result, the light client
+> needs to terminate in case this invariant is violated.
+
+## Supervisor
+
+### Outline
+
+The supervisor implements the functionality of the lightclient. It is
+initialized with a genesis file or with a lightblock the user
+trusts. This initialization is subjective, that is, the security of
+the lightclient is based on the validity of the input. If the genesis
+file or the lightblock deviate from the actual ones on the blockchain,
+the lightclient provides no guarantees.
+
+After initialization, the supervisor awaits an input, that is, the
+height of the next lightblock that should be obtained. Then it
+downloads, verifies, and cross-checks a lightblock, and if all tests
+go through, the light block (and possibly other lightblocks) are added
+to the lightstore, which is returned in an output event to the user.
+
+The following main loop does the interaction with the user (input,
+output) and calls the following two functions:
+
+- `InitLightClient`: it initializes the lightstore either with the
+  provided lightblock or with the lightblock that corresponds to the
+  first block generated by the blockchain (by the validators defined
+  by the genesis file)
+- `VerifyAndDetect`: takes as input a lightstore and a height and
+  returns the updated lightstore.
+
+#### **[LC-FUNC-SUPERVISOR.1]:**
+
+```go
+func Sequential-Supervisor (initdata LCInitData) (Error) {
+
+    lightStore,result := InitLightClient(initData);
+    if result != OK {
+        return result;
+    }
+
+    loop {
+        // get the next height
+        nextHeight := input();
```

Comment:
@cezarad says: in the Rust implementation there are mutliple input actions : VerifytoLatest, LatestTrusted, and GetStatus. The information can be easily obtained from the lightstore. But add a comment for completeness